frontend/src/hocs/withScreencast/withScreencast.tsx

```diff
@@ -0,0 +1,267 @@
+/**
+ * @license
+ * Copyright 2018-2020 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import classNames from "classnames"
+import WhichBrowser from "which-browser"
+import ScreenCastRecorder from "lib/ScreenCastRecorder"
+import hoistNonReactStatics from "hoist-non-react-statics"
+import React, { PureComponent, ComponentType, ReactNode } from "react"
+
+import {
+  UnsupportedBrowserDialog,
+  ScreencastDialog,
+  VideoRecordedDialog,
+} from "hocs/withScreencast/components"
+
+import "./withScreencast.scss"
+
+interface WithScreenCastProps {}
+
+interface WithScreenCastState {
+  fileName: string
+  recording: boolean
+  recordAudio: boolean
+  countdown: number
+  startAnimation: boolean
+  blob?: Blob
+  showRecordedDialog: boolean
+  showScreencastDialog: boolean
+  showUnsupportedDialog: boolean
+}
+
+export interface ScreenCastHOC extends WithScreenCastState {
+  toggleRecordAudio: () => void
+  startRecording: (fileName: string) => void
+  stopRecording: () => void
+}
+
+const SUPPORTED_BROWSERS: any = {
+  Chrome: "49",
+  Firefox: "29",
+  Opera: "36",
+  Edge: "76",
+}
+
+function withScreencast(
+  WrappedComponent: ComponentType<any>
+): ComponentType<any> {
+  class ComponentWithScreencast extends PureComponent<
+    WithScreenCastProps,
+    WithScreenCastState
+  > {
+    static readonly displayName = `withScreencast(${WrappedComponent.displayName ||
+      WrappedComponent.name})`
+
+    recorder?: ScreenCastRecorder | null
+
+    state = {
+      fileName: "streamlit-screencast",
+      countdown: -1,
+      recording: false,
+      recordAudio: false,
+      startAnimation: false,
+      showRecordedDialog: false,
+      showScreencastDialog: false,
+      showUnsupportedDialog: false,
+    }
+
+    toggleRecordAudio = (): void => {
+      const { recordAudio } = this.state
+
+      this.setState({
+        recordAudio: !recordAudio,
+      })
+    }
+
+    showDialog = (fileName: string): void => {
+      const { recording, countdown } = this.state
+
+      if (!this.checkSupportedBrowser()) {
+        this.setState({
+          showUnsupportedDialog: true,
```

Comment:
There is no guarantee here that some other `showFooDialog` boolean isn't also set to `true`. Same with other booleans, below. Please use the state machine approach I mentioned above to solve it.